### PR TITLE
Rule: SC2-4-4-link-has-name

### PR DESCRIPTION
**Pull request by jkodu**
# Checklist for Pull Request
🚨Please review the [guidelines for contributing](https://auto-wcag.github.io/auto-wcag/pages/contribute.html) to this repository.

- [x] Make sure you are requesting to **pull a issue/feature/bugfix branch** (right side). Don't request your master branch!
- [x] Make sure you are making a pull request against the **master branch** (left side). Also you should start *your branch* off *master branch*.
- [x] Kindly prefix the pull request name with `DRAFT` or `FINAL` accordingly. For more details, read [Rule Design Guidelines](https://auto-wcag.github.io/auto-wcag/pages/design/review.html)
- [x] Provide a reference to the issue being addressed.

#Reviewing Guidelines
- For `DRAFT` PR's, leave any comments you have and approve if you’ve reviewed the entire pull request. 
- Once there have been three such reviews the pull request can be merged. 
- Any comments left will be taken up in an updated pull request, which will be necessary before a FINAL release”.
- For `FINAL` PR's, review and approve changes to be merged.

# How to Review And Approve
- Go to the “files changed” tab, there you will have the option to leave comments on different lines. 
- Once the review is completed, find the “Review changes” button in the top right, select “approve” and click “Submit review”

### Description
- Change Draft to Final - https://github.com/auto-wcag/auto-wcag/pull/117
- Close Issue - https://github.com/auto-wcag/auto-wcag/issues/123

💔Thank you!

### CODE REVIEW

**Comment by jkodu** [comment]
@annethyme 
I have to add a few more tests you mentioned in https://github.com/auto-wcag/auto-wcag/pull/117
Pair?

File: _rules/SC2-4-4-link-has-name.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+```
+
+```html
+<!-- Accessible name for link via aria-labelledby -->
+<a href="http://www.w3.org/WAI"><img src="#" aria-labelledby="id1"/></a>
+<div id="id1">This is my link text</div>
+```
+
+```html
+<!-- Passing because this rule doesn't consider whether the accessible name is descriptive enough -->
```
Comment (WilcoFiers) [inline _rules/SC2-4-4-link-has-name.md]: I don't think this is a useful test case. I'd prefer we remove it. Not a deal breaker.

File: _rules/SC2-4-4-link-has-name.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+<!-- Link with icon inserted via font-awesome -->
+<a href="http://www.w3.org/WAI"><i class="fa fa-download"></i></a>
+```
+
+```html
+<!-- Link with image that has empty title -->
+<a href="http://www.w3.org/WAI"><img src="#" title=""/></a>
+```
+
+```html
```
Comment (WilcoFiers) [inline _rules/SC2-4-4-link-has-name.md]: I think we're overdoing it with the img tests here.

File: _rules/SC2-4-4-link-has-name.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+## Background
+
+- [https://www.w3.org/WAI/WCAG20/quickref/?showtechniques=244#navigation-mechanisms-refs](https://www.w3.org/WAI/WCAG20/quickref/?showtechniques=244#navigation-mechanisms-refs)
+- [ARIA7: Using aria-labelledby for link purpose](https://www.w3.org/TR/WCAG20-TECHS/ARIA7.html)
+- [ARIA8: Using aria-label for link purpose](https://www.w3.org/TR/WCAG20-TECHS/ARIA8.html)
+- [F89: Failure of Success Criteria 2.4.4, 2.4.9 and 4.1.2 due to using null alt on an image where the image is the only content in a link](http://www.w3.org/TR/WCAG20-TECHS/F89.html)
+
+## Test Cases
+
+### Passed
```
Comment (WilcoFiers) [inline _rules/SC2-4-4-link-has-name.md]: Empty line after "passed" - also, this needs tests for:
- passing when link is off screen
- passing when link is not in accessibility tree

File: _rules/SC2-4-4-link-has-name.md (hunk 2 above)
Comment (jkodu) [reply _rules/SC2-4-4-link-has-name.md]: No harm in having multiple tests.

File: _rules/SC2-4-4-link-has-name.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+authors:
+- Wilco Fiers
+- Anne Thyme Nørregaard
+---
+
+## Test procedure
+
+### Applicability
+
+The rule applies to any HTML or SVG element with the [semantic role](#semantic-role) of `link` that is [visible on the page](#visible-on-the-page) or [exposed to assistive technologies](#exposed-to-assistive-technologies).
```
Comment (WilcoFiers) [inline _rules/SC2-4-4-link-has-name.md]: There are no SVG tests. I suggest we remove this part and tackle SVG at a later stage.

File: _rules/SC2-4-4-link-has-name.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
+---
+name: Links have an accessible name
+description: |
+  Each link has an accessible name
+
+success_criterion:
+- 4.1.2 # Name, Role, Value
```
Comment (carlosapaduarte) [inline _rules/SC2-4-4-link-has-name.md]: You might want to mention additionally 2.4.4 here

File: pages/algorithms/semantic-role.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
```
Comment (carlosapaduarte) [inline pages/algorithms/semantic-role.md]: This is outdated, compared to the current file

File: _rules/SC2-4-4-link-has-name.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+The rule applies to any HTML element with the [semantic role](#semantic-role) of `link` that is [visible on the page](#visible-on-the-page) or [exposed to assistive technologies](#exposed-to-assistive-technologies).
+
+### Expectation
+
+Each target element has an [accessible name](#accessible-name) that is [non-empty](#non-empty).
+
+## Assumptions
+
+- The rule assumes that all links are user interface components as defined by WCAG 2. When the link role is used incorrectly, this assumption may not be true.
```
Comment (carlosapaduarte) [inline _rules/SC2-4-4-link-has-name.md]: Can we link to the definition of user interface component?

**Comment by jkodu** [comment]
@carlosapaduarte - updated based on your comments. Cheers for the review.

**Pull request merged**